contracts/util/FeeCollector.sol

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.7.4;
+pragma experimental ABIEncoderV2;
+
+import '@openzeppelin/contracts/access/Ownable.sol';
+import '@openzeppelin/contracts/token/ERC20/ERC20.sol';
+
+contract FeeCollector is Ownable {
+    struct Reward {
+        IERC20 token;
+        uint256 amount;
+    }
+
+    struct RawParticipant {
+        address recipient;
+        Reward[] tokens;
+    }
+
+    mapping(address => Reward[]) public rewards;
+
+    event WithdrawReward(address from, address to, uint amount);
+
+    modifier onlyHasRewards {
+        require(rewards[msg.sender].length > 0, 'No rewards have been assigned to address');
+        _;
+    }
+
+    function setRewards(address _target, Reward[] memory _tokens) public onlyOwner {
+        delete rewards[_target];
+        for (uint i=0; i < _tokens.length; i++) {
+            rewards[_target].push(Reward(_tokens[i].token, _tokens[i].amount));
+        }
+    }
+
+    function setRewardsBulk(RawParticipant[] memory _raw) public onlyOwner {
+        for (uint i=0; i < _raw.length; i++) {
+            setRewards(_raw[i].recipient, _raw[i].tokens);
+        }
+    }
+
+    function getRewards() public view onlyHasRewards returns (Reward[] memory) {
+        return rewards[msg.sender];
+    }
+
+    function withdraw(IERC20 _token) external onlyHasRewards {
+        Reward[] memory temp = rewards[msg.sender];
+
+        for (uint i=0; i < temp.length; i++) {
```

Comment:
@Netail This iteration is a bit expensive, better to change your data structure to a 2 dimensional array which you can use to pick the allocated reward uint256. 

Something like this:
```
// set
mapping(address => mapping(address => uint256)) rewards = [];

// get
rewards[_tokenAddress][msg.sender] = balance
```

Reply:
The `uint256` value could of course also be a third array of Reward structs like you have currently, but I guess you will simplify your code quite a bit when grouping rewards by token address.